src/wazuh_modules/syscollector/syscollector_bsd.c

```diff
@@ -63,410 +64,517 @@
+        }
 
-        DIR *dir;
-        struct dirent *version = NULL;
+        char * end = strrchr(de->d_name, '.');
+
+        if (end && !strcmp(end, ".app")) {
 
-        while ((de = readdir(dr)) != NULL) {
+            cJSON * object = sys_parse_pkg(path);
 
-            if (strncmp(de->d_name, ".", 1) == 0) {
+            if (!object) {
+                mtdebug1(WM_SYS_LOGTAG, "Unable to get package information for '%s'", de->d_name);
                 continue;
             }
 
-            cJSON *object = cJSON_CreateObject();
-            cJSON *package = cJSON_CreateObject();
-            cJSON_AddStringToObject(object, "type", "program");
             cJSON_AddNumberToObject(object, "ID", random_id);
             cJSON_AddStringToObject(object, "timestamp", timestamp);
-            cJSON_AddItemToObject(object, "program", package);
-            cJSON_AddStringToObject(package, "format", "pkg");
-            cJSON_AddStringToObject(package, "name", de->d_name);
+            char * string;
+            string = cJSON_PrintUnformatted(object);
+            cJSON_Delete(object);
 
-            snprintf(path, PATH_LENGTH - 1, "%s/%s", HOMEBREW_APPS, de->d_name);
-            cJSON_AddStringToObject(package, "location", path);
-            cJSON_AddStringToObject(package, "source", "homebrew");
+            mtdebug2(WM_SYS_LOGTAG, "Sending '%s'", string);
+            wm_sendmsg(usec, queue_fd, string, LOCATION, SYSCOLLECTOR_MQ);
 
-            dir = opendir(path);
-            if (dir != NULL) {
-                while ((version = readdir(dir)) != NULL) {
-                    if (strncmp(version->d_name, ".", 1) == 0 || strncmp(version->d_name, "..", 2) == 0) {
-                        continue;
-                    }
+            os_free(string);
 
-                    cJSON_AddStringToObject(package, "version", version->d_name);
-                    snprintf(path, PATH_LENGTH - 1, "%s/%s/%s/.brew/%s.rb", HOMEBREW_APPS, de->d_name, version->d_name, de->d_name);
-
-                    char read_buff[OS_MAXSTR];
-                    FILE *fp;
-
-                    if ((fp = fopen(path, "r"))) {
-                        int found = 0;
-                        while(fgets(read_buff, OS_MAXSTR - 1, fp) != NULL && !found){
-                            if (strstr(read_buff, "desc \"") != NULL) {
-                                found = 1;
-                                char ** parts = OS_StrBreak('"', read_buff, 3);
-                                cJSON_AddStringToObject(package, "description", parts[1]);
-                                int i;
-                                for (i = 0; parts[i]; ++i) {
-                                    free(parts[i]);
-                                }
-                                free(parts);
-                            }
-                        }
-                        fclose(fp);
-                    }
-                }
-                closedir(dir);
+        }
+
+        if (recursive) {
```

Comment:
This can be potentially dangerous. I think we should set a limit (a depth of `5` perhaps) to avoid an unlimited number of function calls and opened file descriptors.

Reply:
You are right. I have removed the recursive option due to the results were not good enough.